### PR TITLE
BUG: Fix norm type promotion

### PR DESCRIPTION
**Pull request by toslunar**
Reopen  #10667:

> I tested `np.linalg.norm` with the following code, but it seems the type is not preserved correctly.
> 
> ```python
> import numpy as np
> 
> for dtype in [np.float16, np.float32, np.float64]:
>     a = np.ones((2,), dtype=dtype)
>     ret = np.linalg.norm(a, 3, None, False)
>     if not (dtype == ret.dtype):
>         print('expected', dtype, 'but got', ret.dtype)
> ```
> 
> Without this fix (NumPy 1.14.1):
> 
> ```
> expected <class 'numpy.float16'> but got float64
> expected <class 'numpy.float32'> but got float64
> ```
> 
> With NumPy 1.14.1 + this fix, nothing should be printed.
> 
> Related to #10368 and [cupy/cupy#875 (comment)](https://github.com/cupy/cupy/pull/875#discussion_r170492773)

### CODE REVIEW

**Comment by toslunar** [comment]
Without the bugfix, the new tests fail:
```
% python runtests.py -t numpy/linalg/tests/test_linalg.py
Building, see build.log...
Build OK
NumPy version 1.20.0.dev0+bef7aa0
NumPy relaxed strides checking option: True
NumPy CPU features: SSE SSE2 SSE3 SSSE3* SSE41* POPCNT* SSE42* AVX* F16C* FMA3* AVX2* AVX512F? AVX512CD? AVX512_KNL? AVX512_SKX? AVX512_CLX? AVX512_CNL? AVX512_ICL?

(snip)

FAILED numpy/linalg/tests/test_linalg.py::TestNormDouble::test_vector_return_type - AssertionError:
FAILED numpy/linalg/tests/test_linalg.py::TestNormSingle::test_vector_return_type - AssertionError:
FAILED numpy/linalg/tests/test_linalg.py::TestNormInt64::test_vector_return_type - AssertionError:
3 failed, 290 passed, 1 skipped, 28 deselected, 1 xfailed, 1 warning in 5.12s
```

**Comment by mattip** [comment]
Windows is failing.

**Comment by toslunar** [comment]
Windows is failing at `ord=2`.  Thus, it seems some function in the fast path (maybe `sqnorm = dot(x, x)` or `ret = sqrt(sqnorm)`) unintentionally changes the dtype from `longdouble` to `float64`.

**Comment by toslunar** [comment]
`type(np.sqrt(np.longdouble(2)))` returns `np.float64` on Windows.  I'll relax the test to compare dtypes, instead of comparing `.dtype.type`s.

File: numpy/linalg/tests/test_linalg.py [hunk 1]
```diff
@@ -1221,6 +1221,13 @@ class _TestNormBase:
     dt = None
     dec = None
 
+    @staticmethod
+    def check_dtype(a, res):
+        if issubclass(a.dtype.type, np.inexact):
+            assert_equal(res.dtype, a.real.dtype)
+        else:
+            assert_(issubclass(res.dtype.type, np.floating))
```
Comment (BvB93) [inline numpy/linalg/tests/test_linalg.py:1229]: Note that this `else` branch is redundant here, as `np.floating` is a subclass of `np.inexact`.

**Comment by toslunar** [comment]
It seems the fixed tests `test_vector_return_type`, `test_matrix_return_type` should have been methods of `TestNorm_NonSystematic` because they don't respect `self.dt`.  May I refactor it?

File: numpy/linalg/tests/test_linalg.py [hunk 2]
```diff
@@ -1221,6 +1221,12 @@ class _TestNormBase:
     dt = None
     dec = None
 
+    @staticmethod
+    def check_dtype(a, res):
+        assert_(issubclass(res.dtype.type, np.floating))
+        if issubclass(a.dtype.type, np.inexact):
+            assert_equal(res.dtype, a.real.dtype)
```
Comment (BvB93) [inline numpy/linalg/tests/test_linalg.py:1228]:
```suggestion
        assert_equal(res.dtype, a.real.dtype)
```
To clarify: `a.dtype.type` is _always_ a `np.inexact` subclass if the first assertion passes, so might as well directly skip to the second assert.

I suppose that, at this point, you could also consider removing the dedicated assertion method (again), considering it's only a single line of code.
I'll leave that choice up to you though.

Comment (toslunar) [reply numpy/linalg/tests/test_linalg.py:1228]: The existing code `for each_inexact_types in all_types` is misleading.  It runs over integer types, too.

Comment (BvB93) [reply numpy/linalg/tests/test_linalg.py:1228]: Ah, so that's why that `if` statement is there.
Nevermind in that case.

Comment (mattip) [reply numpy/linalg/tests/test_linalg.py:1228]: A comment to clarify that would be helpful

**Comment by BvB93** [comment]
> It seems the fixed tests `test_vector_return_type`, `test_matrix_return_type` should have been methods of `TestNorm_NonSystematic` because they don't respect `self.dt`. May I refactor it?

By the looks of it `TestNorm_NonSystematic` appears to hold various miscellaneous tests.
Based on this I feel that the current test arrangement is marginally more suitable, despite them ignoring `self.dt`.

File: numpy/linalg/tests/test_linalg.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/linalg/tests/test_linalg.py]: I think this was clearer the way it was before

Comment (BvB93) [reply numpy/linalg/tests/test_linalg.py]: > I think this was clearer the way it was here before @BvB93's suggestion

Yeah, in light of https://github.com/numpy/numpy/pull/17709#discussion_r518008120 I would agree with this.

**Comment by toslunar** [comment]
> > It seems the fixed tests `test_vector_return_type`, `test_matrix_return_type` should have been methods of `TestNorm_NonSystematic` because they don't respect `self.dt`. May I refactor it?
> 
> By the looks of it `TestNorm_NonSystematic` appears to hold various miscellaneous tests.
> Based on this I feel that the current test arrangement is marginally more suitable, despite them ignoring `self.dt`.

That's true.  `TestNorm_NonSystematic` is not the place.  But there was a minor issue: repeating the tests three times in subclasses.  How about a288d826b5a603cbffa7bd1a1849a378c049b8a0?

**Review by eric-wieser — COMMENTED** [review]
I think this probably needs a compatibility release note, but otherwise looks good to me

**Comment by toslunar** [comment]
[This part of the release note of 1.14.0](https://github.com/numpy/numpy/pull/9522/commits/351fe979580ab842cc5c04e1b616d3a64936a695) is for the issue this PR fixes.  `ord in [Inf, -Inf, 0, 1]` (and `[2, None]` with `axis`) were fixed then, but `2.001` was not.

**Comment by seberg** [comment]
We discussed this today at the meeting and I think the general consensus is that this can probably go in.  It may be good to have a quick look over the changes.
In general, it could make sense to do the promotion (i.e. return float64), but that should align with all other code paths (and probably things like `sum` and `prod` if that happens).

**Review by rossbar — APPROVED** [review]
I took the liberty of updating the release note, but everything else LGTM and CI is green (minus circleCI, for unrelated reasons). Thanks @toslunar !

**Comment by toslunar** [comment]
Thank you for the reviews and improving the release note!

**Pull request merged**